### ISSUE TITLE
Showing and focusing window doesn't work properly

### ISSUE THREAD

**Issue by ovalseven8** [qTox/qTox#1985]
Hello,

today I tested the 'Show window' and 'Focus window' functionalities on qTox (latest version from pkg.tox.chat).

I am using Ubuntu Unity 14.04 and unfortunately I have to say that the mentioned features doesn't work properly.

Sometimes it works but most of the time it fails. It's hard for me to describe the problem accurate.
Perhaps someone could test it carefully and comprehensively on different platforms and especially on Ubuntu Unity.

For example 'Show window' only works for me if also 'Focus window' is on. And even with both on it only works sometimes.
The specific difference between 'Show window' and 'Focus window' is also not really clear to me.

Sorry that I am not able to describe the problem more exact. Anyway, I hope that someone can fix it.

Thank you.

**Edit:** Did some further testing -> https://github.com/tux3/qTox/issues/1985#issuecomment-135028937

**Comment by ProMcTagonist** [comment]
> The specific difference between 'Show window' and 'Focus window' is also not really clear to me.

Showing brings the window on top of other windows so it is visible.

Focus means your cursor is blinking in qTox instead of whatever you were just doing.

https://en.wikipedia.org/wiki/Focus_(computing)

**Comment by ovalseven8** [comment]
I've already suspected that. Thanks for the information.

**Comment by ovalseven8** [comment]
@agilob Why got this issue the 'invalid'-label?

The problem still isn't solved.

**Comment by zetok** [comment]
> For example 'Show window' only works for me if also 'Focus window' is on.

.

> Showing brings the window on top of other windows so it is visible.

@ovalseven8 so, you're basically saying that window doesn't appear on top when it should, unless you  have also `Focus window` turned on?

Sorry, but due to the way in which you wrote things, it's very hard to actually understand what your problem is (if there's any problem at all).

If you could, please read https://github.com/tux3/qTox/wiki/Writing-Useful-Bug-Reports

**Comment by ovalseven8** [comment]
@zetok As I wrote in the initial post, yes it's hard to describe the problem. But it the issue is still up-to-date.

Perhaps a developer could test it on Ubuntu Unity and make his own experiences?
@TheSpiritXIII

**Comment by TheSpiritXIII** [comment]
Are you able to build and test #1938? I had to rewrite the window code and I don't remember where but I found some old window activating that the Qt documentation said was an invalid way of doing it.

**Comment by ovalseven8** [comment]
Unfortunately, I don't know how to do that. :/
So, I hope that it will be fixed with your commits, then. :)

But at the moment it just doesn't work correctly.

**Comment by zetok** [comment]
@ovalseven8 I'm reopening issue for now, but if info won't be provided, after some time bug will be simply closed again `:/`

**Comment by ovalseven8** [comment]
> @ovalseven8 so, you're basically saying that window doesn't appear on top when it should, unless you have also Focus window turned on?

Yes, and also if Focus window is on, it doesn't work everytime. Cannot you reproduce it?

**Comment by ovalseven8** [comment]
> Are you able to build and test #1938? I had to rewrite the window code and I don't remember where but I found some old window activating that the Qt documentation said was an invalid way of doing it.

Built qTox with the new multi-window feature. The issue is still not solved. @TheSpiritXIII 
Perhaps you can test it comprehensively on Lubuntu/Ubuntu if you have time.

**Comment by agilob** [comment]
Is https://github.com/tux3/qTox/issues/1542 related?

**Comment by zetok** [comment]
@ovalseven8 

> Yes, and also if Focus window is on, it doesn't work everytime. Cannot you reproduce it?

Well, I don't even really have a DE - I'm using i3wm, which probably doesn't really have stuff DEs have..

@agilob maybe, if it's about focus problems, but this issue isn't about tray.

**Comment by ovalseven8** [comment]
Did some further testings:
- If qTox is minimized by 'Minimize to tray', everything seems to work well.
- If qTox is minimized by 'Close to tray', we have some problems:
  - _No matter whether 'Focus window' is on or off: The qTox window gets focused._ Perhaps someone else can also test it, but for me it seems to be a bug.

---

Additionally, a string I would like to change, because it caused and will cause misunderstandings. 
~~Show window (Show qTox's window when you receive new message.)~~
**Open** window (**Open** qTox's window when you receive **a** new message **and no window is open yet.**) -> https://github.com/tux3/qTox/pull/2124

**Comment by agilob** [comment]
> No matter whether 'Focus window' is on or off: The qTox window gets focused. Perhaps someone else can also test it, but for me it seems to be a bug.

It's related to what I posted, someone introduced that bug and refused to fix it. I'll fix it soon and let me know if that solved problems.

**Comment by PoisonousJohn** [comment]
I'm using version of this commit 1874c0ce4d878f1a179894900431c68f3c751eb9 and still experiencing the problem with focusing

**Comment by vith** [comment]
Just want to confirm what @ovalseven8 described is still happening. If the window exists, it will come to front and steal focus when a message is received even with `Open window` and `Focus window` both unchecked.

This can easily result in sending the end of what you were typing elsewhere as a message in qTox, even a password.

On Arch Linux / gnome-shell with `community/qtox 1.2.4-3`.

**Comment by zetok** [comment]
@vith you have outdated qTox.

**Comment by vith** [comment]
@zetok right you are. I manually rebuilt the arch package for the 1.3.0 tag, same problem there, and also from a48c4225938834db0d4406cff54a07378e9636f4, again same problem.

**Comment by vith** [comment]
Well this is embarrassing. I was stepping through [Widget::newMessageAlert](https://github.com/tux3/qTox/blob/master/src/widget/widget.cpp#L1223-L1250) when I noticed a notification pop up and immediately disappear when the window got focused. Then I checked my gnome extensions and saw I had [Focus my window](https://extensions.gnome.org/extension/1005/focus-my-window/) enabled. Sorry if I wasted your time. It works correctly for me now.

On the plus side, I noticed and fixed a tiny issue with the "Focus window" checkbox: #3048

**Comment by ovalseven8** [comment]
My latest testing results:
- If qTox is minimized by 'Minimize to tray':
  - **Focus window doesn't work!**
  - **The opened window when you receive a new message is resized everytime, what's annoying**
- If qTox is minimized by 'Close to tray':
  - **No matter whether 'Focus window' is on or off: The qTox window gets focused.**
  - At least when the window is opened, it has the same size compared to when I closed it. :smiley:  (Unlike to 'Minimize to tray')
